### DISCUSSION TITLE
Keyboard shortcut plugin for the floating toolbox

### DISCUSSION THREAD

**Discussion by arynvi** [xournalpp/xournalpp#5683] [Q&A]
I asked this in a comment but maybe it is worth to have a space on its own.

I want to write a plugin to bind the floating toolbox to a keyboard shortcut.

I have looked the documentation for plugins: it says it shouldn't be that difficult but I have not been able to do it :(

What I wrote in my plugin is as follows:

```
function initUi()
  app.registerUi({["menu"] = "Open floating toolbox", ["callback"] = "toolbox", ["accelerator"] = "<Shift><Alt>t"});
end

local toggleBox = false

function toolbox()
	if toggleBox == false then
		app.uiAction({["action"] = "ACTION_TOOL_FLOATING_TOOLBOX", ["enabled"] = true});
	else
		app.uiAction({["action"] = "ACTION_TOOL_FLOATING_TOOLBOX", ["enabled"] = false});
	end
	toggleBox = not toggleBox
end
```
But it does not work: first, it opens the toolbox only after I click somewhere, instead of just opening the toolbox; second, I can not choose anything from the toolbox: the only thing I can do is to click another tool to exit the sort of waiting mode it puts itself into. 

I guess I wrote some code really dumb, but I can't figure a way out.

Any help will be appreciated.

Thanks

**Comment by rolandlo** [comment]
Your code looks more or less fine, but the floating toolbox action needs to know where to show up, so you do you have to click somewhere after you press your keyboard accelerator for the plugin action. Moreover you have to use a toolbar with tools in the floating toolbox (the default one doesn't contain any), otherwise it won't work. 

This reminds me that we should add tools to the floating toolbox for all provided toolbars. In the development version (nightly build) tools can be duplicated throughout the toolbars, so there is no reason to keep the floating toolbox empty.

Another question is how to customize the tools in the floating toolbox via the gui. Seems like currently one has to edit the `toolbar.ini` manually.

Your `toolbox` function could be simplified by the way:

```lua
function toolbox()
	toggleBox = not toggleBox
        app.uiAction({["action"] = "ACTION_TOOL_FLOATING_TOOLBOX", ["enabled"] = toggleBox});
end
```

Also I noticed from the code, that `enabled` property is unused for the floating toolbox action. In the stable version when enabled is false, nothing happens. This is a minor inconsistency. The floating toolbox action switches the current tool to the floating toolbox tool (until you click somewhere and choose a tool from the floating toolbox or from one of the toolbars). You don't need a `toggleBox` variable. Just use

```lua
function toolbox()
        app.uiAction({["action"] = "ACTION_TOOL_FLOATING_TOOLBOX"});
end
```

**Reply by tattsan** [reply]
I am just facing the same problem. I use an external program to click the pen position.
``` lua
function invokeFloatingToolbox()
   app.uiAction({["action"]="ACTION_TOOL_FLOATING_TOOLBOX"});
   os.execute("/usr/bin/xdotool click 1 sleep 0.01 key ctrl+shift+p");
end
```
Moreover, I need some actions like  `ctrl+shift+p` to activate the floating toolbox.

**Reply by arynvi** [reply]
@rolandlo 

You are right: the function can be reduced to a single one line instruction and it behaves the same as the original.

As for the other points you make:

- my floating toolbox is populated with tools: I already use it having bound it to the button on my pen. The point is I want to set the button free because I prefer to use it for the eraser, yet I still want to use the floating toolbox somehow and the only other option is via a mouse button, which I find quite uncomfortable.
- As for the necessity of clicking somewhere, it is not so if the toolbox is bound to the pen button: the toolbox shows up by just pressing the button, without the need to click/touch anywhere in the screen, so I do not understand why for a keyboard shortcut it should be any different: I just want the toolbox to show up where the pointer is in that precise moment when I press the keyboard shortcut.
- Finally, in any case it doesn't work: when the floating toolbox shows up after having clicked somewhere, most of the tools in it are greyed out and you can not select them. I include two images that give a hint about that: as you can see, in the "keyboard" image (the first one) the only tools you can select are the two selection tools in the upper right corner and the insert image tool in the lower left corner. If I click anywhere else it doesn't happen anything: the box just disappears and if I click somewhere it shows up again. On the other side, if I use the mouse button (second image) I can select every tool and everything works just fine.

**Reply by arynvi** [reply]
@tattsan I tried the workaround you suggest: it works, in the sense that I do not need to click any longer.
Still, I am not able to select any tool, as described in the other reply.

**Reply by tattsan** [reply]
You may need to unlock `Alt` modifier in your accelerator `<Shift><Alt>t`  like:
```
xdotool keyup Alt_R Alt_L
```
before you activate the floating box (my `'xdotool key ctrl+shift+p'`) .

**Reply by arynvi** [reply]
@tattsan This doesn't change anything.

But thanks to your workaround, I have been able to get it: having bound the floating toolbox to the right click of the mouse, the function

```
function toolbox()
	os.execute("/usr/bin/xdotool click 3 sleep 0.01 key alt+shift+t");
end
```
does the trick for me! :)

So now, at least, I have it functioning. 
It is not the way it should work, but for the moment I can live with that.

**Reply by tattsan** [reply]
> It is not the way it should work, 

I agree.

**Comment by arynvi** [comment]
When I use the keyboard shortcut:
![keyboard](https://github.com/xournalpp/xournalpp/assets/133959396/4bdbc632-3021-4722-9d77-4b7322dcc2c7)

When I use the mouse button:
![mouse_click](https://github.com/xournalpp/xournalpp/assets/133959396/e9659cf5-0a6c-40ff-ac2d-a86b5468c489)

**Comment by bhennion** [comment]
> Another question is how to customize the tools in the floating toolbox via the gui. Seems like currently one has to edit the `toolbar.ini` manually.

If the floating toolbox is activated in the settings, then it shows up in the toolbar customization mode. To activate it in the settings, see Settings -> Drawing Area -> Action on tool tap.

**Comment by youssefadly237** [comment]
Hello everyone,

I’m a medical student using Xournal++ for note-taking, and I’m not deeply familiar with coding (just some python), so I apologize if I lack technical details.

### My Issue:
I want to assign a keyboard shortcut to activate the Floating Toolbox (not to select the tool) at my cursor’s (of the graphic tablet not the mouse) position without changing my currently selected tool. However, when I use:

```lua
app.uiAction({ action = "ACTION_TOOL_FLOATING_TOOLBOX" })
```
it selects it as the active tool, deselecting the tool I am using. (for example I want to change the pen color, but I can not because the pen is getting deselected)

### What I Need:
I just want the Floating Toolbox to appear at my cursor’s position while keeping my current tool active.

### What I Found:
I looked into the source code and found that `showFloatingToolbox()` in `XojPageView` seems to be responsible for displaying the toolbox:

```cpp
void XojPageView::showFloatingToolbox(const PositionInputData& pos) {
    Control* control = xournal->getControl();
    gint wx = 0, wy = 0;
    GtkWidget* widget = xournal->getWidget();
    gtk_widget_translate_coordinates(widget, gtk_widget_get_toplevel(widget), 0, 0, &wx, &wy);
    wx += round_cast<int>(pos.x) + this->getX();
    wy += round_cast<int>(pos.y) + this->getY();
    control->getWindow()->getFloatingToolbox()->show(wx, wy);
}
```
But I have no idea how to call this function from a Lua script.

### My Question:
Is there any way to access `showFloatingToolbox()` from Lua, or is there an alternative method to make the Floating Toolbox appear without changing the active tool? Any help would be greatly appreciated!

![2025-02-05_20-40_1](https://github.com/user-attachments/assets/b39e196e-5445-4c52-9b37-2e0664d0dce3)

**Reply by rolandlo** [reply]
I think there's currently no way to access `showFloatingToolbox(const PositionInputData& pos)` from Lua.  So it takes some changes from the C++ side. We would probably have to split showing the floating toolbox from recording the position where the toolbox should be shown.

Maybe add API so that one can write:
```lua
pos_x, pos_y = app.getCursorPosition()
app.showFloatingToolbox(pos_x, pos_y)
```

**Reply by youssefadly237** [reply]
Thanks for replying!
That makes sense. Having `app.showFloatingToolbox(pos_x, pos_y)` would be perfect. Is there a way I can implement this myself, or does it require deeper changes in the code?

**Reply by rolandlo** [reply]
Well, you can try. It probably doesn't take big changes in the code. You must understand a bit how the communication between Lua and C works, but looking at `luapi_application.h` should suffice. 

The function `showFloatingToolbox(const PositionInputData& pos)` could be turned into `showFloatingToolbox(double x, double y)`, since it doesn't need any more than `pos.x` and `pos.y` currently.

For getting the cursor position you would probably use `gdk_window_get_device_position` for Gtk 3 or `gdk_surface_get_device_position` for Gtk 4.

**Reply by youssefadly237** [reply]
Thanks for the info! I’ll give it a try. If I can't get it working, will this be something that could be included in future updates?

**Reply by rolandlo** [reply]
Well yes, it it works nicely and doesn't disrupt the C++ code base much, it would be merged into master (and be available in the nightly build and any stable Xournal++ version >= 1.3.0).
